### PR TITLE
Finding next mounted step

### PR DESCRIPTION
**Pull request by goloveychuk**
Currently when step is not mounted library skips it, but stays in undefined state.
With this pr it will try to find next mounted step. If any - library will trigger `finished` status.

### CODE REVIEW

**Comment by gilbarbara** [comment]
Can you please check if 1.7.0 stills have this problem and if so update your PR?
Thanks

**Comment by goloveychuk** [comment]
@gilbarbara merged

**Comment by gilbarbara** [comment]
But did you try 1.7.0? @IanVS refactored the step process and maybe it got fixed?

**Comment by goloveychuk** [comment]
I only looked on the code and didn't find any line which can fix it.
will try tomorrow

**Comment by IanVS** [comment]
@goloveychuk can you explain the problem a bit more? How can it be reproduced?

**Comment by goloveychuk** [comment]
just tested, in 1.7.0 it still reproducing. In my fork works ok.
@IanVS let's say you have 3 steps.
First one is mounted in DOM, two others don't.
So lib shows first step. I'm clicking next. Lib tries to find next step - it can't, it writing warning to console, incrementing index. Thats all. Now we have some intermediate state. Walkthough didn't end, steps don't show for user. 
I think correct behavior there is to try find next mounted step. And if any - notify with 'finished' state.

**Comment by IanVS** [comment]
OK thanks, that helps.  I agree this is a problem.

However, I'm not convinced that the correct response is to keep iterating through steps until one is found which can be displayed.  In my mind, not finding a target is a critical problem, and the joyride should not continue past that point by simply skipping steps.  

One possible reason for not finding a target is that the target is animating in to the DOM.  In fact, I face this situation in a few places in my own app.  Clicking a button opens a submenu and advances the joyride simultaneously, but the submenu takes a short time to animate in.  Right now I use a `setTimeout` to delay advancing the joyride, but that feels sloppy.

I propose that a more graceful way to handle a missing target would be to check for the target every X ms for a total of Y ms (something like 16 and 500 respectively), after which if it is still not found, to stop/pause the joyride (similar to sending `joyride.stop()`) while maintaining the step index which was not found.  A callback can then be sent indicating that the step was not found, and it would be up to the application to decide whether to move to another step, try restarting again, or to take some other action. 

What do you think about a solution like that?

**Comment by goloveychuk** [comment]
Actually yes, we have related problem. 
We started walkthough, some steps gone. Next steps should continue when user make some interaction. So basically it's several stage walkthrough. 
And correct behaviour there, for us, is to pause walktrough. 
And your solution covers this case. 
Timeouts will help in case of slow html transitions.
Callback on `not found` will help to decide programmically what to do next. 
If user want to ignore - joyride.next(). if critical - joyride.stop().
> while maintaining the step index
didn't understand there, but on this scenario step index shouldn't be incremented. It should be done by running joyride.next() on cb.

**Comment by IanVS** [comment]
>on this scenario step index shouldn't be incremented.

Yes, that is what I meant.

@gilbarbara, what do you think of my proposed strategy?  If you agree with us, I will work on a PR to implement it.

**Comment by IanVS** [comment]
I'm wondering now if the timeout is a good idea, or whether the callback is enough and the application can retry if it thinks that is the best option.  I can experiment with it both ways and see how it feels, unless someone else has a strong opinion.

**Comment by gilbarbara** [comment]
hey,
I don't know. I ended up creating this library with the idea that it shouldn't handle all the complexity of animations or component transitions and you should only add a step after is ready to display. 

However I know that are situations where it should wait for something to happen before advancing but I still like to keep it simple..

**Comment by goloveychuk** [comment]
I think restarting, intervals should be handled by client, not in lib. For this we should only add callback on failure.

**Comment by goloveychuk** [comment]
so, my proposal.
1) make new callback type - NOT_FOUND instead of increasing index.
2) developer should catch it and decide what he wants - end walkthough, or add delayed repeating. 
I can submit a pr, it's 3 lines change.

**Comment by IanVS** [comment]
I like it.  How about a type of `error:target_not_found`?

I think it should also `stop` the joyride, right?

**Comment by goloveychuk** [comment]
> error:target_not_found
ok
> I think it should also stop the joyride, right?
I don't know is `stop` clearing state. 
It should `pause` walkthough, that developer can continue from pervious state (step).

**Comment by IanVS** [comment]
Stop is similar to a pause, it just makes the beacon not show up.  The index is not changed, that is done from `reset`.

**Comment by goloveychuk** [comment]
yes, and `isRunning` should became `false`

**Comment by goloveychuk** [comment]
and what should be 
`shouldRedraw`, and `shouldRenderTooltip` ?

**Comment by IanVS** [comment]
Don't setState directly, just call `this.stop()`

**Comment by goloveychuk** [comment]
@IanVS made force push

File: src/scripts/Joyride.jsx [hunk 1]
```diff
@@ -268,6 +269,18 @@ class Joyride extends React.Component {
...
     if (nextState.index !== index && isRunning) {
+      if (nextStep && !this.getStepTargetElement(nextStep)) {
+        console.warn('Target not mounted', nextStep, action); //eslint-disable-line no-console
+        this.triggerCallback({
+          action,
+          index,
+          type: callbackTypes.TARGET_NOT_FOUND,
+          step
+        });
+        this.stop();
```
Comment (IanVS) [inline src/scripts/Joyride.jsx]: I had forgotten that this part of code is inside the `componentWillUpdate`, during which you cannot call `setState`, so stopping does not work. We'll have to do it somewhere else.

**Comment by IanVS** [comment]
@goloveychuk I had to re-work things slightly, and so I opened my own PR that used some of the ideas you had in this PR.  I hope you don't mind, but I'm going to close this PR now.  I would really appreciate if you could review https://github.com/gilbarbara/react-joyride/pull/195, since you're the one who originally brought up this issue and made the first attempt at a fix.

**Pull request closed without merging**